### PR TITLE
Copernicus Atmosphere Data Store - retrieve data from CADS

### PR DESCRIPTION
**Pull request by annefou**
This tool is similar to c3s but for the Copernicus Atmosphere Data Store.

- [x] Create python, xml and bash tool
- [x] Allow user to specify the geographical area to download
- [x] Add test where one file is retrieved
- [x] Add test where geographical area is specified and several files are retrieved, unzip, concatenated and converted to netCDF.

### CODE REVIEW

File: tools/cads/cads.xml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+       #end if
+        ]]></configfile>
+    </configfiles>
+    <inputs>
+        <conditional name="is_file">
+            <param name="has_req" type="select" label="Provide CADS request">
+                <option value="no">paste as text</option>
+                <option value="yes" selected="true">give a filename</option>
+            </param>
+	   <when value="yes">
```
Comment (bgruening) [inline tools/cads/cads.xml:48]: indentation

File: tools/cads/cads.xml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    </inputs>
+    <outputs>
+        <data name="request" format="txt" from_work_dir="request.txt"/>
+        <data name="ofilename" format="netcdf" from_work_dir="tmp.nc"/>
+    </outputs>
+    <tests>
+        <test>
+            <conditional name="is_file">
+                <param name="has_req" value="yes" />
+	        <param name="api_req_file" ftype="txt" value="req.txt" />
```
Comment (bgruening) [inline tools/cads/cads.xml:64]: indentation

File: tools/cads/cads_retrieve.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+print("req = ", req)
+c3s_type = req.split('c.retrieve')[1].split('(')[1].split(',')[0].strip(' "\'\t\r\n')
+
+c3s_req = '{' + req.split('{', 1)[1].rsplit('}', 1)[0].replace('\n', '') + '}'
+c3s_req_dict = ast.literal_eval(c3s_req)
+
+c3s_output = req.rsplit('}', 1)[1].split(',')[1].split(')')[0].strip(' "\'\t\r\n')
+
+f = open(args.output, "w")
```
Comment (bgruening) [inline tools/cads/cads_retrieve.py:42]: with open(args.output, "w") as f:

File: tools/cads/cads_retrieve.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+f.write("dataset to retrieve: " + c3s_type + "\n")
+f.write("request: " + c3s_req + "\n")
```
Comment (bgruening) [inline tools/cads/cads_retrieve.py:44]: use f-strings?

f.write(f'dataset to retrieve: {c3s_type}\nrequest: {c3s_req}\noutput filename: {c3s_output}')

File: tools/cads/cads_retrieve.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+import cdsapi
+
+parser = argparse.ArgumentParser()
+parser.add_argument("-i", "--request", type=str, help="input API request")
+parser.add_argument("-o", "--output", type=str, help="output API request")
+args = parser.parse_args()
+
+if path.isfile(args.request):
+    f = open(args.request, "r")
```
Comment (bgruening) [inline tools/cads/cads_retrieve.py:13]: with open(args.request) as f:

File: tools/cads/cads.xml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+       export HOME=`pwd`  &&
+       #set $cads_apikey = $__user__.extra_preferences.get('cads_account|cads_apikey', "")
+       #if $cads_apikey == ""
+	    echo "Error. Set your CADS credentials via: User -> Preferences -> Manage Information"  &&
+            echo 'Will try to use tool adsapirc file (used for testing)' &&
+            cp '$__tool_directory__/adsapirc.cfg' .cdsapirc &&
+       #else
+            cp '$cds_key_file' .cdsapirc &&
```
[the remark this answers is not in the document]
Comment (annefou) [reply tools/cads/cads.xml:18]: Yes! Thanks!

File: tools/cads/cads_retrieve.py (hunk 4 above)
Comment (annefou) [reply tools/cads/cads_retrieve.py:44]: Cool! I am not very familiar with f-strings but this is really nice! Thanks.

**Review by bgruening — APPROVED** [review]
Look good to me!

**Comment by annefou** [comment]
Thank you!

**Pull request merged**